### PR TITLE
[IMP] Form builder + asynchronous post function Ready to merge

### PR DESCRIPTION
**Pull request by mdo-odoo**
New feature : Website Form Builder
https://www.odoo.com/web#id=9220&view_type=form&model=project.task&menu_id=3942&action=327

### CODE REVIEW

**Comment by KangOl** [comment]
Please rebase and squash your commits

File: addons/website_crm/controllers/main.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# -*- coding: utf-8 -*-
+ # -*- coding: utf-8 -*-
```
Comment (KangOl) [inline addons/website_crm/controllers/main.py]: remove leading space.

File: addons/website_crm/controllers/__init__.py [hunk 2]
```diff
@@ -1 +1 @@
-import main
+import main
```
Comment (KangOl) [inline addons/website_crm/controllers/__init__.py]: keep newline at end of file

File: addons/website_crm/__init__.py [hunk 3]
```diff
@@ -1 +1 @@
-import controllers
+import controllers
```
Comment (KangOl) [inline addons/website_crm/__init__.py]: keep newline at end of file

File: addons/website_crm/controllers/main.py [hunk 4]
```diff
@@ -15,98 +15,4 @@ def generate_google_map_url(self, street, city, city_zip, country_name):
...
-                    'res_name': field_value.filename,
-                    'res_model': 'crm.lead',
-                    'res_id': lead_id,
-                    'datas': base64.encodestring(field_value.read()),
-                    'datas_fname': field_value.filename,
-                }
-                request.registry['ir.attachment'].create(request.cr, SUPERUSER_ID, attachment_value, context=request.context)
-
-        return self.get_contactus_response(values, kwargs)
+        return url
```
Comment (KangOl) [inline addons/website_crm/controllers/main.py]: keep newline at end of file

File: addons/website_form_builder/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    'name': 'Website Form Builder',
+    'category': 'Website',
+    'summary': 'Build your contact form and link it to the database',
+    'version': '1.0',
+    'description': """
+OpenERP Contact Form
+====================
+
+        """,
+    'author': 'OpenERP SA',
```
Comment (KangOl) [inline addons/website_form_builder/__openerp__.py]: Odoo SA

File: addons/website_form_builder/controllers/main.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        model = request.session.get('form_builder_model')
+        record = request.env.sudo()[model].browse(request.session.get('form_builder_id'))
+        if not record.exists():
+            return False
+        return request.website.render(template, {"record":record})
+
+    @http.route('/website_form/<model>', type='http', auth="public", website=True)
+    def website_form(self, model, **kwargs):
+
+        form = request.env['website.form'].search([('model_name', '=', model)])
```
Comment (KangOl) [inline addons/website_form_builder/controllers/main.py]: What if you there is more than one form?

Comment (mdo-odoo) [inline addons/website_form_builder/controllers/main.py]: if you need several form, you dont need to send its on the same time with same parameters. This part dont need to be managed on the controller.

File: addons/website_form_builder/controllers/main.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        except ValueError:
+            return request.redirect('/page/'+template)
+
+        model = request.session.get('form_builder_model')
+        record = request.env.sudo()[model].browse(request.session.get('form_builder_id'))
+        if not record.exists():
+            return False
+        return request.website.render(template, {"record":record})
+
+    @http.route('/website_form/<model>', type='http', auth="public", website=True)
```
Comment (KangOl) [inline addons/website_form_builder/controllers/main.py]: POST only

File: addons/website_form_builder/models/website_form.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
+# -*- coding: utf-8 -*-
+
+import base64
+from openerp import tools
+from openerp.http import request
+from openerp import models, fields, api, _
+
+class website_form(models.Model):
+    """ Manage models, blacklists, etc.. for form builder"""
+    _name = 'website.form'
```
Comment (KangOl) [inline addons/website_form_builder/models/website_form.py]: In fact, you must _inherit `ir.model`

File: addons/website_form_builder/models/website_form.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    def boolean(self, label, input):
+        return (input != 0)
+    
+    def integer(self, label, input):
+        return int(self.get_int(input))
+    
+    def float(self, label, input):
+        return float(self.get_float(input))
+
+    _filter = {'char': char, 'text': text, 'html': html, 'many2one': many2one,
```
Comment (KangOl) [inline addons/website_form_builder/models/website_form.py]: Maybe, you have to re-use ir.qweb fields

File: addons/website_form_builder/models/website_form.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        data['error'] = list(set(field for field in self.get_required() if not data['post'].get(field)))
+        return data
+    
+    def insert_message(self, data, id_record):
+        if self.model_name == 'mail.mail':
+            return True
+        if not data['message']['attachment_ids']:
+            return True
+        data['message']['res_id'] = id_record
+        return self.env['mail.message'].sudo().create(data['message']).id
```
Comment (KangOl) [inline addons/website_form_builder/models/website_form.py]: use `post_message` on the model ?

**Comment by KangOl** [comment]
And fix runbot errors

**Comment by rim-odoo** [comment]
@mdo-odoo Can you add a link to your task please ?

**Comment by mdo-odoo** [comment]
I added the link on my first comment. I will rebase and fix tests issues monday 27 april because i m working on an other task now.

**Comment by rim-odoo** [comment]
Thx

**Pull request closed without merging**